torch/csrc/jit/custom_operator.h

```diff
@@ -0,0 +1,190 @@
+  // If there is no '(' in the schema, we assume this is only the name (e.g.
+  // "foo::bar").
+  const auto bracketIndex = schemaOrName.find('(');
+  if (bracketIndex == std::string::npos) {
+    // Infer the full schema and we're good.
+    return torch::jit::detail::createFunctionSchemaFromTraits<Traits>(
+        /*name=*/schemaOrName);
+  }
+
+  // If the user provided her own schema, we need to infer it nevertheless and
+  // check that it's correct. We return the user provided schema in the end
+  // because it has proper argument names.
+
+  auto providedSchema = parseSchema(schemaOrName);
+
+  const auto inferredSchema =
+      torch::jit::detail::createFunctionSchemaFromTraits<Traits>(
+          providedSchema.name);
+  checkArgumentVector(
+      "argument",
+      inferredSchema.arguments,
+      providedSchema.arguments,
+      inferredSchema,
+      providedSchema);
+  checkArgumentVector(
+      "return value",
+      inferredSchema.returns,
+      providedSchema.returns,
+      inferredSchema,
+      providedSchema);
+  return providedSchema;
+}
+} // namespace detail
+
+/// Registers a custom operator with a name or schema, and an implementation
+/// function.
+///
+/// If the first argument specifies only the function name like `foo::bar`, the
+/// schema, including the type of each argument and the return type, is inferred
+/// from the function signature. Otherwise, the string should specify the whole
+/// schema, like `foo::bar(Tensor a, double b) -> Tensor`. In that case, the
+/// schema will still be inferred from the function and checked against this
+/// provided schema.
+///
+/// If the schema is left to be inferred, the argument names will take on
+/// sequential placeholder names like `_0`, `_1`, '_2' and so on. If you want
+/// argument names to be preserved, you should provide the schema yourself.
+///
+/// The implementation function can be a function pointer or a functor
+/// (including a lambda object). The function (or `operator()`) can take any
+/// number of arguments with a type from the subset accepted by the PyTorch
+/// JIT/Script backend, and return a single type or a tuple of types.
+///
+/// Example invocation:
+/// ```
+/// createOperator(
+///    "foo::bar(float a, Tensor b)",
+///    [](float a, at::Tensor b) { return a + b; });
+/// ```
+template <typename Implementation>
+Operator createOperator(
+    const std::string& schemaOrName,
+    Implementation&& implementation) {
+  using Traits = c10::guts::infer_function_traits_t<Implementation>;
+  using ArgumentTypes =
+      c10::guts::typelist::map_t<decay_t, typename Traits::parameter_types>;
+  using ArgumentTuple =
+      typename c10::guts::typelist::to_tuple<ArgumentTypes>::type;
+  using ReturnType = decay_t<typename Traits::return_type>;
+
+  auto schema = torch::jit::detail::inferAndCheckSchema<Traits>(schemaOrName);
+
+  return Operator(schema, [implementation](Stack& stack) {
```

Comment:
style nit: Can we use pointers for mutable arguments? i.e. `Stack*` instead of `Stack&`?

Reply:
We don't do this anywhere in PyTorch or ATen -- that's the Google Style Guide?